File: 2020/python/day03.py
import unittest


class Grid:
    def __init__(self, columns, rows):
        self.data = set()
        self.rows = rows
        self.columns = columns

    def add(self, entry):
        self.data.add(entry)


def parse(grid: Grid, y, line):
    x = 0
    for c in line:
        if c == '#':
            grid.add((x, y))
        x = x + 1


def step(grid: Grid, position):
    x, y = position
    return (x, y) in grid.data


def update_position(columns, position, step_x=3, step_y=1):
    x, y = position
    return ((x + step_x) % columns, y + step_y)


def run(grid, step_x=3, step_y=1):
    position = (0, 0)
    hits = 0
    while position[1] < grid.rows:
        position = update_position(grid.columns, position, step_x, step_y)
        if step(grid, position):
            hits = hits + 1
    return hits


def part1():
    grid = read_input()
    print(run(grid))


def part2():
    grid = read_input()
    print(
        run(grid, 1, 1) * run(grid) * run(grid, 5, 1) * run(grid, 7, 1) *
        run(grid, 1, 2))


def read_input():
    with open('../inputs/input03') as input:
        inputs = [i.strip() for i in input if not i.isspace()]
        grid = Grid(len(inputs[0]), len(inputs))
        y = 0
        for i in inputs:
            parse(grid, y, i)
            y = y + 1
        return grid


class TestPart1(unittest.TestCase):
    def test_parse(self):
        grid = Grid(11, 1)
        parse(grid, 0, '..##.......')
        self.assertEqual(grid.data, {(2, 0), (3, 0)})

    def test_step(self):
        grid = Grid(11, 3)
        grid.data = {(2, 0), (3, 0), (0, 1), (4, 1), (8, 1), (1, 2), (6, 2),
                     (9, 2)}
        self.assertEqual(step(grid, (6, 2)), True)

    def test_update_position(self):
        self.assertEqual(update_position(11, (10, 0)), (2, 1))
        self.assertEqual(update_position(11, (0, 0)), (3, 1))
        self.assertEqual(update_position(11, (8, 0)), (0, 1))

    def test_run(self):
        grid = Grid(11, 5)
        grid.data = {(2, 0), (3, 0), (0, 1), (4, 1), (8, 1), (1, 2), (6, 2),
                     (9, 2), (2, 3), (4, 3), (8, 3), (10, 3), (1, 4), (5, 4),
                     (6, 4), (9, 4)}
        self.assertEqual(run(grid), 2)


class TestPart2(unittest.TestCase):
    def test_run(self):
        grid = Grid(11, 5)
        grid.data = {(2, 0), (3, 0), (0, 1), (4, 1), (8, 1), (1, 2), (6, 2),
                     (9, 2), (2, 3), (4, 3), (8, 3), (10, 3), (1, 4), (5, 4),
                     (6, 4), (9, 4)}
        self.assertEqual(run(grid, 1, 2), 1)
        self.assertEqual(run(grid, 7, 1), 2)
        pass


if __name__ == '__main__':
    part1()
    part2()
    unittest.main()

File: 2020/cpp/day03.cpp
#include <cassert>
#include <fstream>
#include <iostream>
#include <set>
#include <tuple>
#include <vector>

typedef std::tuple<int, int> Position;
typedef std::tuple<int, int> Size;

void grid_from_input(const std::vector<std::string> &lines,
                     std::set<Position> &positions, Size &size) {
  int x = 0;
  int y = 0;
  for (auto line : lines) {
    x = 0;
    for (auto c : line) {
      if (c == '#')
        positions.insert(Position(x, y));
      x++;
    }
    y++;
  }
  size = {x, y};
}

bool hit(const std::set<Position> &grid, const Position &position) {
  return grid.find(position) != grid.end();
}

void update_position(const int max_x, Position &position, const int step_x,
                     const int step_y) {
  auto [x, y] = position;
  position = {(x + step_x) % max_x, y + step_y};
}

void read_input(std::vector<std::string> &lines) {
  std::ifstream in("../inputs/input03");
  std::string str;
  while (std::getline(in, str)) {
    if (str.find_first_not_of(' ') != std::string::npos) {
      lines.push_back(str);
    }
  }
}

int run(const std::set<Position> &grid, const Size &size, const int step_x,
        const int step_y) {
  auto [columns, rows] = size;
  Position position = {0, 0};
  int hits = 0;
  while (std::get<1>(position) < rows) {
    update_position(columns, position, step_x, step_y);
    auto [x, y] = position;
    if (hit(grid, position))
      hits++;
  }
  return hits;
}

void part1() {
  std::vector<std::string> lines;
  std::set<Position> grid;
  Size size;
  read_input(lines);
  grid_from_input(lines, grid, size);
  std::cout << run(grid, size, 3, 1) << std::endl;
}

void part2() {
  std::vector<std::string> lines;
  std::set<Position> grid;
  Size size;
  read_input(lines);
  grid_from_input(lines, grid, size);
  std::cout << run(grid, size, 1, 1) * run(grid, size, 3, 1) *
                   run(grid, size, 5, 1) * run(grid, size, 7, 1) *
                   run(grid, size, 1, 2)

            << std::endl;
}

void test_grid_from_input() {
  std::vector<std::string> lines{"..##......."};
  std::set<Position> positions;
  Size size;
  grid_from_input(lines, positions, size);
  assert(positions.size() == 2);
  auto [columns, rows] = size;
  assert(columns == 11);
  assert(rows == 1);
}

void test_hit() {
  std::set<Position> positions{{2, 0}, {3, 0}};
  assert(!hit(positions, {1, 0}));
  assert(hit(positions, {2, 0}));
  assert(!hit(positions, {2, 1}));
}

int main(void) {
  test_grid_from_input();
  test_hit();
  part1();
  part2();
  return 0;
}
